File: 03-tobogganTrajectory/main.py
input_list = open("input.txt").read().strip().split("\n")

forest = [[char for char in line] for line in input_list]

tree_char = "#"

def find_collisions(forest:list, x_offset:int, y_offset:int) -> int:
    encountered_trees = 0
    x_pointer = 0

    for row in forest[::y_offset]:
        target_index = x_pointer % len(row)
        if row[target_index] == tree_char:
            encountered_trees += 1
        x_pointer += x_offset

    return encountered_trees

offset_pairs = [
    (3, 1),
    (1, 1),
    (5, 1),
    (7, 1),
    (1, 2)
]

tree_product = 1

for i, pair in enumerate(offset_pairs):
    encountered_trees = find_collisions(forest, *pair)
    tree_product *= encountered_trees
    print(f"Pair {pair}: {encountered_trees} trees", ("(part one solution)" if i == 0 else ""))

print(f"Product of all: {tree_product} (part two solution)")

File: 03-tobogganTrajectory/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

var tree_char = []rune("#")[0] // No idea why I can't just do rune("#")

func main() {
	inputBytes, _ := ioutil.ReadFile("input.txt")
	inputSlice := strings.Split(strings.TrimSpace(string(inputBytes)), "\n")

	var forest [][]rune
	for _, line := range inputSlice {
		forest = append(forest, []rune(line))
	}

	offsetPairs := [][]int{
		{3, 1},
		{1, 1},
		{5, 1},
		{7, 1},
		{1, 2},
	}

	treeProduct := 1

	for i, pair := range offsetPairs {
		encounteredTrees := findCollisions(forest, pair[0], pair[1])
		treeProduct *= encounteredTrees
		fmt.Printf("Pair %d: %d trees", i, encounteredTrees)
		if i == 0 {
			fmt.Print(" (part one solution)")
		}
		fmt.Println()
	}

	fmt.Printf("Product of all: %d (part two solution)\n", treeProduct)

}

func findCollisions(forest [][]rune, xOffset, yOffset int) int {
	var encounteredTrees int
	var xPointer int
	var yPointer int

	for yPointer < len(forest) {
		row := forest[yPointer]
		targetIndex := xPointer % len(row)
		if row[targetIndex] == tree_char {
			encounteredTrees += 1
		}
		xPointer += xOffset
		yPointer += yOffset
	}

	return encounteredTrees
}
